Components: Draggable, add story

ItsJonQ opened the pull request:
This update adds a Storybook example for the Draggable component from `@wordpress/components`.

![Screen Capture on 2019-10-22 at 15-06-44](https://user-images.githubusercontent.com/2322354/67321911-3627b600-f4de-11e9-9bd7-c7e3ef38d1d0.gif)

Adding another one to:
https://github.com/WordPress/gutenberg/issues/17973

❤️

Commit:
Components: Add Story for Draggable

This update adds a Storybook example for the Draggable component from `@wordpress/components`.

gziolo commented on packages/components/src/draggable/stories/index.js:
```diff
@@ -0,0 +1,65 @@
+};
+
+export const _default = () => {
+	const [ isDragging, setDragging ] = useState( false );
```
According to documentation, the function exported as a story isn't a React component. It means that we shouldn't use hooks in this context as they might stop to work in the near future if they change the way all that works.

The easiest way to get rid of this warning is to move the part which manages state out of this function. See a similar usage:
https://github.com/WordPress/gutenberg/blob/a6e1252886e8d990627ac4bf4796f768bfad2a44/packages/components/src/checkbox-control/stories/index.js#L18-L28

In this case, it might be more work to do it :(

Commit:
Merge branch 'master' into add/draggable-story

Commit:
Fix useState hook for Draggable story example

Solution was to create an Example component with the useState hook.
Render that Example component in the story instead.

ItsJonQ commented:
@gziolo Ah! Thank you so much 🤗 

After creating a couple of other stories, I discovered that creating a dedicated `Example` component with the hook stuff resolves this issue:

```jsx
const Example = () => (...)

export const _default = () => <Example />
```

I just updated this story with that setup, and it looks like it's working 👍

gziolo commented on packages/components/src/draggable/stories/index.js:
```diff
@@ -0,0 +1,69 @@
+
+export default { title: 'Draggable', component: Draggable };
+
+const Box = ( props ) => {
```
Yet another reason to introduce a Box component which would work cross-platform :)

gziolo commented:
> After creating a couple of other stories, I discovered that creating a dedicated `Example` component with the hook stuff resolves this issue:
> 
> ```js-jsx
> const Example = () => (...)
> 
> export const _default = () => <Example />
> ```
> 
> I just updated this story with that setup, and it looks like it's working 👍

Yes, this should work. Sometimes you can extract only bits that handle the state and connect it with the original component but in this case using `Example` is the most straightforward way.

gziolo commented on packages/components/src/draggable/stories/index.js:
```diff
@@ -0,0 +1,69 @@
+	return (
+		<div>
+			<p>Is Dragging? { isDragging ? 'Yes' : 'No' }</p>
+			<div id="draggable-example-box" style={ { display: 'inline-flex' } }>
```
I'm a bit concerned about the implementation of the `Drabbable` component. Why do you need to wrap it with a `div` which has the `id` which is referenced with `elementId` prop?

It's completely unrelated to this PR, but I'd be in favor of opening a follow-up issue which makes it easier to use.

Taking the example from README:

```jsx
import { Dashicon, Draggable, Panel, PanelBody } from '@wordpress/components';

const MyDraggable = ( { onDragStart, onDragEnd } ) => (
	<div id="draggable-panel">
		<Panel header="Draggable panel" >
			<PanelBody>
				<Draggable
					elementId="draggable-panel"
					transferData={ { } }
					onDragStart={ onDragStart }
					onDragEnd={ onDragEnd }
				>
				{
					( { onDraggableStart, onDraggableEnd } ) => (
						<Dashicon
							icon="move"
							onDragStart={ onDraggableStart }
							onDragEnd={ onDraggableEnd }
							draggable
						/>
					)
				}
				</Draggable>
			</PanelBody>
		</Panel>
	</div>
);
```

I would love us to explore whether it could be refactored to:

```jsx
import { Dashicon, Draggable, Panel, PanelBody } from '@wordpress/components';

const MyDraggable = ( { onDragStart, onDragEnd } ) => (
	<Draggable
		as={ Panel }
		header="Draggable panel"
		transferData={ { } }
		onDragStart={ onDragStart }
		onDragEnd={ onDragEnd }
	>
		{ ( { onDraggableStart, onDraggableEnd } ) => (
			<PanelBody>
				<Dashicon
					icon="move"
					onDragStart={ onDraggableStart }
					onDragEnd={ onDraggableEnd }
					draggable
				/>
			</PanelBody>
		) }
	</Draggable>
);
```

/cc @nosolosw

gziolo replied:
There is also a related issue with draggable elements as explained in:
https://github.com/WordPress/gutenberg/pull/18158

ItsJonQ replied:
> Why do you need to wrap it with a div which has the id which is referenced with elementId prop?

@gziolo I noticed that too. I found it a bit unintuitive/restrictive in my first couple attempts of getting `Draggable` to work.

The above `as` technique is cool! And if `as` is not defined, maybe it can fallback to a regular `<div>`, wrapping the `children` that has been passed in.

gziolo reviewed (approved):
Let's get this in. It works great in my testing. I left an unrelated comment in regards to the implementation details of the component.
